lib/compat/wordpress-6.2/get-global-styles-and-settings.php

```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * API to interact with global settings & styles.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Function to get the settings resulting of merging core, theme, and user data.
+ *
+ * @param array $path    Path to the specific setting to retrieve. Optional.
+ *                       If empty, will return all settings.
+ * @param array $context {
+ *     Metadata to know where to retrieve the $path from. Optional.
+ *
+ *     @type string $block_name Which block to retrieve the settings from.
+ *                              If empty, it'll return the settings for the global context.
+ *     @type string $origin     Which origin to take data from.
+ *                              Valid values are 'all' (core, theme, and user) or 'base' (core and theme).
+ *                              If empty or unknown, 'all' is used.
+ * }
+ *
+ * @return array The settings to retrieve.
+ */
+function gutenberg_get_global_settings( $path = array(), $context = array() ) {
+	if ( ! empty( $context['block_name'] ) ) {
+		$path = array_merge( array( 'blocks', $context['block_name'] ), $path );
+	}
+	$origin = 'custom';
+	if ( isset( $context['origin'] ) && 'base' === $context['origin'] ) {
+		$origin = 'theme';
+	}
+
+	// TODO: offer a way to clean cache, if neccessary.
+	static $settings_by_origin = array(
```

Comment:
The only change here is to add the `$settings_by_origin` cache. The rest is moving the function from 6.0 compat folder to 6.2 compat folder.

Reply:
I came here for the object cache too. 👍 It'll never be as fast as static vars, but it should be close.

Reply:
If you use a persistent object cache, then the second page load, will avoid all processing.

Reply:
> If you use a persistent object cache

I'd worry about over-caching here. I'm not sure we can confidently control all the variables that determine the final settings object, especially with filters around. Unless we are very confident about how we invalidate a persistent cache — which probably requires some dexterity around filter output — I'd avoid it.

Reply:
For now, we could use `wp_cache_add_non_persistent_groups` to add a non persistent group, until we workout cache invalidation.

Reply:
How confident are we that `wp_cache_add_non_persistent_groups` is well supported? The docs [say](https://developer.wordpress.org/reference/classes/wp_object_cache/#wp_cache-functions) not all plugins use it. I've checked the four plugins referenced in the page (memcached, redis, w3 total cache, litespeed) and litespeed does not use it as far as I could find.

Reply:
I've also started to compile some events that should [invalidate the cache](https://github.com/WordPress/gutenberg/pull/45372#issuecomment-1296859561). Some are necessary for persistent caching across requests, others are relevant for non-persistent cache as well (static or object).

Reply:
Lightspeed does seem to support it https://plugins.trac.wordpress.org/browser/litespeed-cache/trunk/src/object.lib.php#L133

Reply:
Oh, interesting. Thanks for bringing that up. There must be some error when I checked the code locally, as I don't have the src folder (and many other things). Going to check what happened:

![image](https://user-images.githubusercontent.com/583546/199696395-bb0e08e4-69d8-4645-bf6a-80a9e9919feb.png)

Reply:
I must have checked out a particular svn revision. Did a full checkout and everything looks good.